File: LeetCodeJS/CourseSchedule_207.js
// courseSchedule.js
class CourseSchedule_207 {
    static UNVISITED = 0;
    static VISITING = 1;
    static VISITED = 2;

    canFinish(numCourses, prerequisites) {
        const graph = new Map();

        // Build the graph
        prerequisites.forEach(([course, prereq]) => {
            if (!graph.has(course)) {
                graph.set(course, []);
            }
            graph.get(course).push(prereq);
        });

        const states = Array(numCourses).fill(CourseSchedule_207.UNVISITED);

        const dfs = (course) => {
            if (states[course] === CourseSchedule_207.VISITING) return false; // Cycle detected
            if (states[course] === CourseSchedule_207.VISITED) return true;   // Already visited

            states[course] = CourseSchedule_207.VISITING;
            const neighbors = graph.get(course) || [];

            for (let neighbor of neighbors) {
                if (!dfs(neighbor)) {
                    return false;
                }
            }

            states[course] = CourseSchedule_207.VISITED;
            return true;
        };

        for (let i = 0; i < numCourses; i++) {
            if (!dfs(i)) {
                return false;
            }
        }

        return true;
    }
}

// Export the CourseSchedule class
module.exports = CourseSchedule_207;

File: LeetCodeKotlin/CourseSchedule_207.kt
class CourseSchedule_207 {
    private enum class Color { UNVISITED, VISITING, VISITED }

    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        /* 1. build the graph */
        val graph = Array(numCourses) { mutableListOf<Int>() }
        for (p in prerequisites) graph[p[0]].add(p[1])      // edge: course -> prereq

        /* 2. colour array */
        val color = Array(numCourses) { Color.UNVISITED }

        /* 3. DFS */
        fun dfs(u: Int): Boolean {
            when (color[u]) {
                Color.VISITING -> return false              // back‑edge  ⇒ cycle
                Color.VISITED  -> return true               // already safe
                Color.UNVISITED -> {}                       // fall through
            }
            color[u] = Color.VISITING
            for (v in graph[u])
                if (!dfs(v)) return false
            color[u] = Color.VISITED
            return true
        }

        /* 4. explore every component */
        for (i in 0 until numCourses)
            if (!dfs(i)) return false
        return true
    }
}
